### PR TITLE
Completion provides invalid suggestions for xsd:sequence

### LINKED ISSUE #347: Completion provides invalid suggestions for xsd:sequence
```xml
<?xml version="1.0" encoding="utf-8"?>
<xs:schema xmlns:xs="http://www.w3.org/2001/XMLSchema" elementFormDefault="qualified" attributeFormDefault="unqualified">
    <xs:element name="data">
        <xs:complexType>
            <xs:sequence>
                <xs:element name="e1"/>
                <xs:element name="e2"/>
            </xs:sequence>
        </xs:complexType>
    </xs:element>
</xs:schema>
```
Given the schema above the code completion shows the following behavior:

- for the very fist element in `<data>` the code completion suggests both elements `<e1>` and `<e2>`. But only `<e1>` is a valid choice.
- for the second element in <data> the code completion suggests both elements` <e1>` and `<e2>`. But only `<e2>` is a valid choice.

The schema validation correctly points out the invalid elements. But code completion should not suggest invalid elements in the first place.

I observed this behavior with redhat.vscode-xml v0.5.0

### PR DESCRIPTION
**Pull request by angelozerr**
Fix #347

Signed-off-by: azerr <azerr@redhat.com>

### CODE REVIEW

File: org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/extensions/contentmodel/model/CMDocument.java [hunk 1]
```diff
@@ -13,6 +13,7 @@
 import java.util.Collection;
 
 import org.eclipse.lsp4xml.dom.DOMElement;
+import org.w3c.dom.Element;
```
Comment (fbricon) [inline org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/extensions/contentmodel/model/CMDocument.java]: unnecessary import

Comment (angelozerr) [reply org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/extensions/contentmodel/model/CMDocument.java]: @fbricon updated

**Comment by fbricon** [comment]
doesn't work for tag.xsd
```xml
<?xml version="1.0" encoding="UTF-8" ?>
<root
    xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
    xsi:noNamespaceSchemaLocation="tag.xsd">
  <tag></tag>
  <|
</root>
```
at the cursor position, `tag` is proposed (I think it shouldn't), not `optional`.

**Comment by angelozerr** [comment]
@fbricon there were a bug with find of sibling element with a given offset. It should be fixed now. I have add a lot of tests with that.

File: org.eclipse.lsp4xml/src/test/java/org/eclipse/lsp4xml/extensions/contentmodel/XMLSchemaCompletionExtensionsTest.java [hunk 2]
```diff
@@ -259,13 +281,13 @@ public void completionWithoutStartBracket() throws BadLocationException {
 				"<beans xmlns=\"http://www.springframework.org/schema/beans\" xsi:schemaLocation=\"http://www.springframework.org/schema/beans http://www.springframework.org/schema/beans/spring-beans.xsd\" xmlns:xsi=\"http://www.w3.org/2001/XMLSchema-instance\">\r\n"
 				+ //
 				"	|";
-		testCompletionFor(xml, c("bean", "<bean></bean>"));
+		//testCompletionFor(xml, c("bean", "<bean></bean>"));
```
[the remark this answers is not in the document]
Comment (angelozerr) [reply org.eclipse.lsp4xml/src/test/java/org/eclipse/lsp4xml/extensions/contentmodel/XMLSchemaCompletionExtensionsTest.java]: @fbricon updated

File: org.eclipse.lsp4xml/src/test/java/org/eclipse/lsp4xml/extensions/contentmodel/XMLSchemaCompletionExtensionsTest.java [hunk 3]
```diff
@@ -259,13 +281,13 @@ public void completionWithoutStartBracket() throws BadLocationException {
...
 				"	|";
-		testCompletionFor(xml, c("bean", "<bean></bean>"));
+		//testCompletionFor(xml, c("bean", "<bean></bean>"));
 
 		xml = "<?xml version=\"1.0\" encoding=\"UTF-8\"?>\r\n" + //
 				"<beans xmlns=\"http://www.springframework.org/schema/beans\" xsi:schemaLocation=\"http://www.springframework.org/schema/beans http://www.springframework.org/schema/beans/spring-beans.xsd\" xmlns:xsi=\"http://www.w3.org/2001/XMLSchema-instance\">\r\n"
 				+ //
 				"	|" + "</beans>";
-		testCompletionFor(xml, c("bean", "<bean></bean>"));
+		//testCompletionFor(xml, c("bean", "<bean></bean>"));
```
[the remark this answers is not in the document]
Comment (angelozerr) [reply org.eclipse.lsp4xml/src/test/java/org/eclipse/lsp4xml/extensions/contentmodel/XMLSchemaCompletionExtensionsTest.java]: @fbricon updated

File: org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/services/AbstractPositionRequest.java [hunk 4]
```diff
@@ -52,21 +52,47 @@ public DOMNode getNode() {
...
+			DOMElement element = (DOMElement) currentNode;
+			if (element.isClosed() && currentNode.getEnd() <= offset) {
+				return element;
+			}
+		}
+		return currentNode.getPreviousSiblingElement();
+	}
+
+	/**
+	 * DOMNode previousSibling = request.getPreviousSiblingElement();//
```
[the remark this answers is not in the document]
Comment (angelozerr) [reply org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/services/AbstractPositionRequest.java]: Please retry again @fbricon I have written test with your usecases.It should work.

**Comment by fbricon** [comment]
Found some issues with choices. Using this schema:
```xml
<?xml version="1.0" encoding="utf-8" ?>
<xs:schema
        xmlns:xs="http://www.w3.org/2001/XMLSchema"
        elementFormDefault="qualified"
        attributeFormDefault="unqualified">
    <xs:element name="person">
        <xs:complexType>
            <xs:choice
                    minOccurs="1"
                    maxOccurs="3">
                <xs:element name="employee"/>
                <xs:element name="member"/>
            </xs:choice>
        </xs:complexType>
    </xs:element>
</xs:schema>
```
Then:
```xml
<?xml version="1.0" encoding="UTF-8" ?>
<person
        xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
        xsi:noNamespaceSchemaLocation="choice.xsd">
    <employee></employee>
    <member></member>
    <| // No more completion (should have either employee or member
</person>
```
and then
```xml
<?xml version="1.0" encoding="UTF-8" ?>
<person
        xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
        xsi:noNamespaceSchemaLocation="choice.xsd">
    <employee></employee>
    <employee></employee>
       ... // add any amount of employees > 3
    <employee></employee>
    <| returns both employee and member, shouldn't return anything
</person>
```
=> as soon as both elements have been added and number of siblings > 3, then no more completions are proposed

**Comment by angelozerr** [comment]
Good catch @fbricon ! because of your usecase, I had to change my code which consumes Xerces validator and now I think I have a very clean code and less complex. I have add a test with your use case.

**Comment by fbricon** [comment]
Works nicely now. Great job @angelozerr !

**Comment by angelozerr** [comment]
Thank's @fbricon! I think we cover all usescases with tests and since we use Xerces validator `XSCMValidator`  which is used too in the XML Schema validator, completion which returns possible elements should be OK now.

**Pull request merged**